### PR TITLE
Count xxx

### PR DESCRIPTION
**Pull request by jtlap**

### CODE REVIEW

File: include/eve/module/core/regular/countl_one.hpp [hunk 1]
```diff
@@ -7,10 +7,24 @@
...
+#include <eve/module/core/decorator/core.hpp>
+#include <eve/module/core/regular/countl_zero.hpp>
+#include <bit>
 
 namespace eve
 {
+  template<typename Options>
+  struct countl_one_t : elementwise_callable<countl_one_t, Options>
+  {
+    template<eve::value T>
```
Comment (jfalcou) [inline include/eve/module/core/regular/countl_one.hpp:21]: All functions must be on `eve::unsigned_value`

File: include/eve/module/core/regular/countr_zero.hpp [hunk 2]
```diff
@@ -48,10 +61,60 @@ namespace eve
...
 //================================================================================================
-EVE_MAKE_CALLABLE(countr_zero_, countr_zero);
-}
+  inline constexpr auto countr_zero = functor<countr_zero_t>;
 
-#include <eve/module/core/regular/impl/countr_zero.hpp>
+  namespace detail
+  {
+    template<value T, callable_options O>
+    EVE_FORCEINLINE constexpr T countr_zero_(EVE_REQUIRES(cpu_),
```
Comment (jfalcou) [inline include/eve/module/core/regular/countr_zero.hpp:69]: put it back into `impl/countr_zero.hpp`

File: include/eve/module/core/regular/countr_zero.hpp [hunk 3]
```diff
@@ -48,10 +61,60 @@ namespace eve
...
+        using elt_t        = element_type_t<T>;
+        x &= eve::inc(~x);
+        T c(sizeof(elt_t) * 8 - 1);
+        if constexpr( siz == 8 )
+        {
+          c -= if_else(is_nez(x & T(0x0F)), T(4), zero);
+          c -= if_else(is_nez(x & T(0x33)), T(2), zero);
+          c -= if_else(is_nez(x & T(0x55)), T(1), zero);
+        }
+        else if constexpr( siz <= 16 )
```
Comment (jfalcou) [inline include/eve/module/core/regular/countr_zero.hpp:88]: == 16

File: include/eve/module/core/regular/impl/simd/arm/neon/countl_zero.hpp [hunk 4]
```diff
@@ -14,19 +14,20 @@
...
-  else if constexpr( c == category::uint8x8 ) return vclz_u8(a0);
-  else return map(countl_zero, a0);
-}
+    if      constexpr( c == category::uint32x4 ) return vclzq_u32(a0);
+    else if constexpr( c == category::uint32x2 ) return vclz_u32(a0);
+    else if constexpr( c == category::uint16x8 ) return vclzq_u16(a0);
+    else if constexpr( c == category::uint16x4 ) return vclz_u16(a0);
+    else if constexpr( c == category::uint8x16 ) return vclzq_u8(a0);
+    else if constexpr( c == category::uint8x8  ) return vclz_u8(a0);
+    else return map(countl_zero, a0);
```
Comment (jfalcou) [inline include/eve/module/core/regular/impl/simd/arm/neon/countl_zero.hpp:31]: `countl_zero.behavior(cpu{}...`

File: include/eve/module/core/regular/impl/simd/arm/sve/countr_zero.hpp [hunk 5]
```diff
@@ -13,29 +13,25 @@
...
-{
-  if constexpr( C::is_complete && C::is_inverted )
+  template<unsigned_scalar_value T, typename N, callable_options O>
+  EVE_FORCEINLINE wide<T, N> countr_zero_(EVE_REQUIRES(sve_),
+                                          O          const&,
+                                          wide<T, N> const& v) noexcept
+  requires sve_abi<abi_t<T, N>>
   {
-    return countl_zero(wide<T,N>(svrbit_x(detail::sve_true<T>(),v)));
+    return countl_zero(svrbit_x(sve_true<T>(),v)));
```
Comment (jfalcou) [inline include/eve/module/core/regular/impl/simd/arm/sve/countr_zero.hpp:22]: should be :
```c++
return countl_zero(wide<T,N>(svrbit_x(sve_true<T>(),v))));
```

**Review by jfalcou — CHANGES_REQUESTED** [review]

File: include/eve/module/core/regular/countl_one.hpp (hunk 1 above)
Comment (jtlap) [reply include/eve/module/core/regular/countl_one.hpp:21]: done

File: include/eve/module/core/regular/countr_zero.hpp (hunk 2 above)
Comment (jtlap) [reply include/eve/module/core/regular/countr_zero.hpp:69]: done

File: include/eve/module/core/regular/countr_zero.hpp (hunk 3 above)
Comment (jtlap) [reply include/eve/module/core/regular/countr_zero.hpp:88]: done

File: include/eve/module/core/regular/impl/simd/arm/neon/countl_zero.hpp (hunk 4 above)
Comment (jtlap) [reply include/eve/module/core/regular/impl/simd/arm/neon/countl_zero.hpp:31]: done

File: include/eve/module/core/regular/impl/simd/arm/sve/countr_zero.hpp (hunk 5 above)
Comment (jtlap) [reply include/eve/module/core/regular/impl/simd/arm/sve/countr_zero.hpp:22]: done

File: include/eve/module/core/regular/impl/simd/arm/neon/countl_zero.hpp (hunk 4 above)
Comment (jtlap) [reply include/eve/module/core/regular/impl/simd/arm/neon/countl_zero.hpp:31]: undone

File: include/eve/module/core/regular/impl/countr_zero.hpp [hunk 6]
```diff
@@ -15,66 +15,55 @@
...
+    if constexpr( scalar_value<T> )
+      return T(std::countr_zero(x));
+    else
     {
-      c -= if_else(is_nez(x & T(0x0000FFFF)), T(16), zero);
-      c -= if_else(is_nez(x & T(0x00FF00FF)), T(8), zero);
-      c -= if_else(is_nez(x & T(0x0F0F0F0F)), T(4), zero);
-      c -= if_else(is_nez(x & T(0x33333333)), T(2), zero);
-      c -= if_else(is_nez(x & T(0x55555555)), T(1), zero);
+      //      return map(countr_zero, x); // TO DO
```
Comment (jfalcou) [inline include/eve/module/core/regular/impl/countr_zero.hpp:31]: remove the TODO I guess ?

**Review by jfalcou — APPROVED** [review]

**Pull request merged**